### DISCUSSION TITLE
Shared worker feedback & support

### DISCUSSION THREAD

**Discussion by novemberborn** [avajs/ava#2703] [Ideas]
We're pretty excited about [shared workers](https://github.com/avajs/ava/blob/master/docs/recipes/shared-workers.md)! But we need more real-world experience in building AVA plugins before we can make it generally available. Please give feedback and build plugins. We'd be more than happy to promote them.

Not sure what to build? Previously folks have expressed a desire for mutexes, managing Puppeteer instances, starting (database) servers and so forth.

We could also extend the shared worker implementation in AVA itself. Perhaps so you can run code before a new test run, even with watch mode. Or so you can initialize a shared worker based on the AVA configuration, not when a test file runs.

Please comment here with ideas, questions and feedback.

**Comment by jakobrosenberg** [comment]
Sharing use-case here by request 🙂

I'm working on a public template where I need to test both `dev` and `build` commands.

I have a folder structure as such
```
test/
  build/
  common/
  dev/
```
and the `package.json` tests
```json
"test": "npm run test:dev && npm run test:build",
"test:dev": "node test/wrapper dev ava test/{common,dev}/**/*.test.js",
"test:build": "node test/wrapper build ava test/{common,build}/**/*.test.js"
```

The `test/wrapper` script
```javascript
// assign `dev` or `build` to mode and `ava` to cmd
const [mode, cmd, ...args] = process.argv.slice(2)

async function wrapSpawn() {
    const teardown = await setups[mode]()
    spawnSync(npx, [cmd, ...args], { stdio: 'inherit' })
    teardown()
}
```

<details>

<summary>entire wrapper script</summary>

```javascript
const { spawnSync, spawn } = require('child_process')
const { checkPort, wait } = require('./utils')
const fkill = require('fkill')

const npm = /^win/.test(process.platform) ? 'npm.cmd' : 'npm'
const npx = /^win/.test(process.platform) ? 'npx.cmd' : 'npx'

// assign `dev` or `build` to mode and `ava` to cmd
const [mode, cmd, ...args] = process.argv.slice(2)

const setups = {
    async dev() {
        const child = await spawn(npm, ['run', 'dev'], )
        await checkPort(5000, 15000)
        await wait(500)
        return () => fkill(child.pid, { tree: true, force: true })
    },
    async build() {
        spawnSync(npm, ['run', 'build'], )
        const child = await spawn(npm, ['run', 'serve'], )
        await checkPort(5000, 15000)
        return () => fkill(child.pid, { tree: true, force: true })
    }
}

async function wrapSpawn() {
    const teardown = await setups[mode]()
    spawnSync(npx, [cmd, ...args], { stdio: 'inherit' })
    teardown()
}

wrapSpawn()
```

</details>


I'm somewhat happy with the script in its current state. It's not 100% self explanatory in `package.json`, but works flawlessly.

---

I had previously used shared context and being able to use `reserve` was great for spinning up a single server. However I couldn't get the teardowns working. Either the server shut down along with the test file that spawned it or it didn't shut down at all.

---

For my use case the best option would be an argument that lets me provide a setup script. (I guess I could DIY that for now)
```
ava test/dev/*.js --setup test/_dev-setup.js
ava test/build/*.js --setup test/_build-setup.js
```

The setup file:
```javascript
//test/_dev-setup.js
module.exports = async function() {
  const handle = await createMyDevServer()
  
  // return a teardown function
  return function(){
    handle.kill()
  }
}
```

**Reply by djake** [reply]
Hi! Did anything ever come of this?

**Comment by novemberborn** [comment]
@jakobrosenberg so the way I was thinking this might work:

* We add support in AVA config files for loading plugins
* We allow plugins to register tasks that run before and after test runs
* (And they'd be aware of watchers, to perhaps keep resources alive for longer)
* The message passing capabilities already available would allow you to load the plugin in a test file and access URLs and whatnot

What do you think?

**Comment by jakobrosenberg** [comment]
Sounds good.

What are your thoughts on being able to override the plugins through the CLI (`npx ava path/to/test --plugin path/to/plugin.js`) and/or dynamic higher order components? (Am I using this term correctly?)

```javascript
//myplugin.js
module.exports = payload => {
  payload.foo = 'bar'
  return process.argv.includes('--my-custom-flag')
    ? require('./alt-plugin')(payload)
    : require('./main-plugin')(payload)
}
  ```

**Comment by novemberborn** [comment]
I'm reluctant to add CLI options for things that you would apply on every test run, I'd rather only expose flags for the stuff you need to do occasionally to debug a test run.

Passing arguments to plugins that have been registered through the configuration could be interesting though.

**Comment by nathanforce** [comment]
Not too sure if this is the right place for this or if a new issue would be better, but I'll start here!

I'll start this all of with a bug caveat that I don't have much experience with workers, memory sharing, or any of the other nitty gritty stuff involved here.

My ultimate goal is to create a `jsdom` instance for each test in my frontend application and avoid [polluting the global namespace](https://github.com/jsdom/jsdom/wiki/Don't-stuff-jsdom-globals-onto-the-Node-global). I want to do this so that I can benefit from Ava's concurrency.

To run my app in `jsdom` I'm bundling the whole thing up using `esbuild`.  When a `<script>` on the page requests my app bundle `jsdom` will respond with a `Buffer` created from `esbuild`'s `OutputFile.contents`, a `Uint8Array`. 

https://github.com/evanw/esbuild/blob/4e336e4f313a1f8a4bd6032f037d23f725f52e75/lib/types.ts#L101

This all works nicely and, importantly, it avoids serializing the contents of the JS file.

This works fine but has a scaling issue as each test suite/process needs to build the app. I wondered if workers might help solve this by building once and providing the output to all tests that need it.  

I've put together a rudimentary example using workers that does indeed reduce the runs of `esbuild` to just 1, but the test time is far slower than multiple runs.  My guess is that this is because the communication between the workers is serializing the build output.

This is where my limited knowledge of this type of thing starts to fade.  It seems [based on the docs for postMessage](https://nodejs.org/docs/latest/api/worker_threads.html#worker_threads_port_postmessage_value_transferlist) that the expected solution here would be to use `SharedArrayBuffer` to avoid copying the data. Would it be possible to expose that API in Ava's shared worker API or is it purposefully not exposed?

Sorry for the long winded backstory! Thanks for all your work.

**Comment by novemberborn** [comment]
@nathanforce if you could repost that as a [Discussion](https://github.com/avajs/ava/discussions) that would be more appropriate. Happy to explore this with you.

**Comment by djake** [comment]
`NODE_ENV` is set to `test` when running tests but not set when running a shared worker.

I'm a little torn on what I think the behavior should be, though as a baseline suggestion I think the behavior should be documented.

It seems like maybe the default should be NODE_ENV="test" for consistency, but whatever the default, I think this should be customizable (without having to write to process.env from within the worker). This would probably mean adding a property to the SharedWorker constructor, which creates a problem similar to the existing initialData property.

**Reply by novemberborn** [reply]
Nice observation! Currently the worker thread inherits the `NODE_ENV` of the main AVA process, which does not care about the environment it's in. If you set `NODE_ENV` when invoking AVA then that should be inherited by the worker thread.

I'm not convinced the worker thread environment should be manipulatable by the `environmentVariables` configuration. Those are meant for the test workers.

One could also argue that `NODE_ENV=test` is only relevant for the code being tested, not for shared worker code.

Could you elaborate on your use case?

**Reply by djake** [reply]
I am doing http endpoint testing. I spin up a postgres docker container with dockerode and launch supertest/express in before hooks, then kill the container in an after.always. The application, including some singletons, as well as the test setup code are configured using [node-config](npmjs.com/package/config) and rely on NODE_ENV to select configuration from an appropriate yaml file, including things like the db name. This is a new small project and all the tests are in one file (yay macros!), and share state via the database. This is fine for starters but I thought it would be a nice opportunity for me to tryout sharedWorkers by spinning up multiple containers or multiple dbs in the one container and then letting each file request a set of db credentials from the worker.

For starters, I just moved my existing setup code from a before hook to a worker. Things stopped working which ultimately turned out to be because the worker was using config from development.yml and the tests were using config from test.yml.

> One could also argue that NODE_ENV=test is only relevant for the code being tested, not for shared worker code.

I don't necessarily disagree with this, which is why the only suggestion I'm confident of is that it be documented. I offered adding a property to the SharedWorker constructor as a potential solution because I agree with your comment that environmentVariables should be reserved for the test workers.

I think the use case for the existing `environmentVariables` is probably something like a) I can do `npx ava` without having to worry about a dotenv file or configuring env vars in my shell and/or b) ava doesn't willy nilly pass along other environment vars to test workers. That's just a guess, and I agree the sharedWorkers and testWorkers shouldn't necessarily share config, but I think the same logic applies to sharedWorkers as the test workers.

Conversely, I think my current option for configuring a sharedWorker's env vars would be: 
* to actually set the environment variable in my shell. That is worse dx (no more simple `npx ava`) and could provoke confusion, e.g. if you want `MY_DB_CREDS=something npx ava` with ava configuration `environmentVariables: { MY_DB_CREDS: 'somethingelse' }`, that's a recipe for a very annoying first experience with ava for a future developer.
* pass some config via initialData and write it to process.env. I have a strong aversion to such direct access of process.env (which is why I use node-config in the first place) and tend to [lint it](https://github.com/mysticatea/eslint-plugin-node/blob/master/docs/rules/no-process-env.md).

Hope that helps. Happy to elaborate further, and thanks for your consideration!

**Reply by novemberborn** [reply]
> For starters, I just moved my existing setup code from a before hook to a worker. Things stopped working which ultimately turned out to be because the worker was using config from development.yml and the tests were using config from test.yml.

Yea that's not good.

> I think the use case for the existing `environmentVariables` is probably something like a) I can do `npx ava` without having to worry about a dotenv file or configuring env vars in my shell and/or b) ava doesn't willy nilly pass along other environment vars to test workers.

You're right on `a`. But all environment variables are passed to the test worker, so `b` doesn't hold.

> I agree the sharedWorkers and testWorkers shouldn't necessarily share config, but I think the same logic applies to sharedWorkers as the test workers.

That's another view. Any worker, test or shared, should have the same environment. So `NODE_ENV` defaults to `'test'`, and `environmentVariables` are applied.

This would be the most consistent approach. What do you think?

**Reply by djake** [reply]
Consistent sounds good to me!

**Comment by timdp** [comment]
I've been experimenting with running Puppeteer in a shared worker (and having it connect to Fastify in another one). It's been smooth sailing so far but there's one thing I can't figure out: robust disposal.

The Node docs mention that there's no `process.on(...)` inside worker threads. As a result, I can't use that to kill the browser on exit, Ctrl+C, etc.

What I've done so far is use the message bus to have workers accept a dispose command. I trigger that command from the teardown hook on the outside. However, if the message bus itself gets stuck waiting for a reply, or if the process gets killed, that doesn't do anything since the message is never received.

In my case, if disposal doesn't run, I get a headless Chrome process that I need to kill manually. Similarly, I use ngrok to add SSL to my Fastify server, and that also spawns a process that doesn't get killed automatically.

It'd be great if AVA itself would let me register a (guaranteed) teardown function _inside_ a worker as well. As far as I can tell, that's not available from AVA today and I also wouldn't know what underlying API to use.

Any guidance around this? Thanks!

**Reply by novemberborn** [reply]
There is no direct "AVA teardown" API at the moment, no. I'd love to model the test run lifecycle rather than just the test worker lifecycle.

For now though… did you notice you can register a teardown callback for each test worker? If you count the number of active workers then after the least teardown you can spin down the processes.

The test worker "instance" is available on received messages. You can also use `protocol.testWorkers()` to get an async iterator for each connecting test worker.

**Reply by timdp** [reply]
> For now though… did you notice you can register a teardown callback for each test worker? If you count the number of active workers then after the least teardown you can spin down the processes.

I think that's what I'm doing. In the test process:

```js
registerSharedWorker({
  filename,
  supportedProtocols: ['experimental'],
  teardown: async () => {
    // using my utility function:
    await sendCommand(worker, '_dispose')
  }
})
```

and then in the worker, something like:

```js
const main = negotiateProtocol(['experimental'])
// ... set up browser here ...
await main.ready()
// using another utility function:
await receiveCommands(main, {
  ...commands,
  _dispose: async () => {
    // ... tear down browser here ...
  }
})
```

I'm not sure why I'd need to count the number of active workers with this approach, since every worker is instantiated exactly once. I guess you were talking about abstracting out the logic that calls the disposal command? As you can see, I chose to do that by defining the `_dispose` action, but that small amount of boilerplate code is indeed repeated between the workers now.

But crucially, is there even an API in Node's worker threads implementation that would allow this? Without `process.on(...)` or an alternative to it, a worker will never be able to detect that it's being terminated. To what extent can AVA itself catch the various exit events and forward them to workers? They're synchronous, so will workers even receive them in time?

_Edit:_ Sharing my utility functions [as a Gist](https://gist.github.com/timdp/97ce4bfcc2ea43ff746e369173af52ec) in case people want to save some time.

**Reply by novemberborn** [reply]
Ideally we expose lifecycle "events" for each test run. We could sketch that out a bit, see who's interested in contributing? This should deal with SIGINT as well.

The `beforeExit` event should fire in the worker thread though, which may give you enough time to stop the child processes.

Additionally you should be able to communicate with the shared worker during a teardown, however to ensure the message is received on the other end you'll have to wait for a reply.

But! The shared worker can register its own teardown for when a test worker exits. That way you don't need to communicate anything from inside the test worker. This is what I was referring to in my earlier post.

P.S. You shouldn't need to serialize errors. Under the hood this uses the [serialization API](https://nodejs.org/docs/latest-v12.x/api/v8.html#v8_serialization_api).

**Reply by timdp** [reply]
> The `beforeExit` event should fire in the worker thread though, which may give you enough time to stop the child processes.

I tried listening for `beforeExit` and `exit` inside a Shared Worker but they didn't seem to fire. I initially thought that was because the docs say:

> Signals are not delivered through `process.on('...')`.

but that clearly only talks about signals. Maybe I did something wrong.

The docs _also_ say:

> All messages sent from the worker thread are emitted before the `'exit'` event is emitted on the Worker object.

so that's a relief. I might wire it up myself using the underlying API.

> But! The shared worker can register its own teardown for when a test worker exits. That way you don't need to communicate anything from inside the test worker. This is what I was referring to in my earlier post.

I only now realized that you were talking about [this section](https://github.com/avajs/ava/blob/eef1a55584307533b6b82b17e8edde47b5eb9fb3/docs/recipes/shared-workers.md#cleaning-up-resources), sorry.

Based on that, why not add a `main.teardown(asyncFn)` that registers a hook for when the thread is about to exit? I still think the challenging part is going to be making it async though.

It'd probably be good to also add some sort of diagram that explains the lifecycle. I've just discovered that my approach with a disposal message doesn't work with multiple test files - the first file sets up the worker and kills it again (which wasn't the idea), so when the second file kicks in, it's already gone. I'm sure I'll figure it out by going through the docs again but I expect others to run into similar issues.

> P.S. You shouldn't need to serialize errors. Under the hood this uses the [serialization API](https://nodejs.org/docs/latest-v12.x/api/v8.html#v8_serialization_api).

Thanks, good to know!

**Reply by timdp** [reply]
With the counts, I assume you meant something like this:

```js
let count = 0
for await (const testWorker of main.testWorkers()) {
  ++count
  testWorker.teardown(async () => {
    if (--count === 0) {
      await dispose()
    }
  })
}
```

That _does_ get the job done with default settings, but with `--serial`, the counter reaches zero after the first test file...

_Edit:_ Dumped my code into a [repo](https://github.com/timdp/ava-shared-workers-experiment) in case you want to test it yourself.

**Reply by novemberborn** [reply]
> I tried listening for `beforeExit` and `exit` inside a Shared Worker but they didn't seem to fire.

I haven't tried this in a clean setup, but I _was_ debugging something yesterday where AVA (inside a worker thread) was handling a `beforeExit`, so it does seem like it should work.

> The docs _also_ say:
> 
> > All messages sent from the worker thread are emitted before the `'exit'` event is emitted on the Worker object.
> 
> so that's a relief. I might wire it up myself using the underlying API.

You don't have access to the `Worker` object though. And presumably this applies only to messages sent via the `parentPort`. With AVA 4 there'll be direct message ports between the shared and test workers (which themselves run inside a worker thread), so they likely won't flush as expected.

> Based on that, why not add a `main.teardown(asyncFn)` that registers a hook for when the thread is about to exit? I still think the challenging part is going to be making it async though.

Yes, hence my comment around the "test run" lifecycle (we'd want this to be usable within watch mode as well).

> It'd probably be good to also add some sort of diagram that explains the lifecycle.

That's a good idea.

**Reply by novemberborn** [reply]
I think in my case `beforeExit` was triggered because the worker thread ran out of work to do. It might not be triggered when it's closed explicitly. Luckily AVA is in charge of that so we can wire up a little protocol for async teardown of the thread itself.

**Reply by novemberborn** [reply]
I have some feedback, after looking at your code.

Generally speaking I think shared workers are best for managing "expensive" resources (or coordinating access to resources across test workers). Disregarding `ngrok`, typically starting and managing a web server inside the test worker is cheap and easier than doing it across a shared worker.

It's Puppeteer that's the expensive resource. However it's likely easier to model what you need it to do, and use the message bus to achieve that. Sketching this out, we can launch Puppeteer when we first need to open a page, and clean it up when there are no test workers left. And pages can be closed based on a simple reply.

You're right that this doesn't work with `--serial`, but if you have resource constraints on the number of pages you want to open, the shared worker should handle these. Tests can wait until their page has been loaded.

```js
const puppeteer = require('puppeteer-core')

module.exports = async ({ negotiateProtocol }) => {
  const protocol = negotiateProtocol(['experimental'])
  protocol.ready()

  // Note: Multiple invocations of `openPage()` may modify this variable.
  let launching = null

  const openPage = async message => {
    if (launching === null) {
      launching = puppeteer.launch()
    }

    // Track for teardown purposes.
    const promise = launching
    
    let tornDown = false
    message.testWorker.teardown(async () => {
      tornDown = true
     
      activeTestWorkers.delete(message.testWorker)
      if (activeTestWorkers.size === 0 && launching === promise) {
        launching = null
        const browser = await promise
        await browser.close()
      }
    })

    const browser = await launching
    if (tornDown) return // Race conditions are fun
    
    const page = await browser.newPage()
    if (tornDown) return // Race conditions are fun
    
    await page.goto(message.data.url)
    if (tornDown) return // Race conditions are fun
    
    for await (const _ of message.reply().replies()) {
      if (tornDown) return // Race conditions are fun

      await page.close()
      return
    }
  }

  for await (const message of protocol.subscribe()) {
    if (message.data.type === 'openPage') {
      openPage(message)
    }
  }
}
```

**Reply by timdp** [reply]
Thanks, I appreciate the feedback!

> It's Puppeteer that's the expensive resource. 

It depends on what you consider "expensive". I know Shared Workers are intended for compute-intensive tasks, but they're also a good fit for a challenge that isn't immediately apparent from my proof of concept: I want to share a single web server between tests so:

1. it can sit behind a single ngrok tunnel because those are limited in number (and therefore expensive in terms of cost), and
2. it can share state if necessary - e.g., if the server needs to fetch external fixtures.

> You're right that this doesn't work with `--serial`, but if you have resource constraints on the number of pages you want to open, the shared worker should handle these.

It's more for debugging purposes that I need `--serial` to work. In case of failures, I know I can also use `--match` but sometimes, it's easier to just run tests one by one and look at verbose logging or set breakpoints.

Anyway, I don't think this addresses the core issue. I think it should be feasible to make this work from the AVA side - e.g., via a `main.teardown(hook)`. I don't think it's a crazy thing to support, given that:

1. there's already a lot of functionality that addresses (in my opinion) less common use cases, and
2. AVA lets you provide what's essentially a constructor for the Shared Worker - the part that gets called before `ready()` - so why not support a destructor as well?

> Tests can wait until their page has been loaded.

Indeed, and they're doing that in my version as well. I don't think that's the concern?

**Reply by novemberborn** [reply]
> I know Shared Workers are intended for compute-intensive tasks, but they're also a good fit for a challenge that isn't immediately apparent from my proof of concept

Your use cases are great! Tunnels are a limited resource, and shared workers should be a great way to manage them. Same for accessing external fixtures. Hopefully tests are not compute heavy, so that typical worker thread use case does not apply here.

It's just there's a trade-off between managing the shared worker and duplicating effort within a single test file. 

> Anyway, I don't think this addresses the core issue. I think it should be feasible to make this work from the AVA side - e.g., via a `main.teardown(hook)`. I don't think it's a crazy thing to support, given that:
> 
> 1. there's already a lot of functionality that addresses (in my opinion) less common use cases, and
> 2. AVA lets you provide what's essentially a constructor for the Shared Worker - the part that gets called before `ready()` - so why not support a destructor as well?

I agree! The functionality that's in place now is what I needed to solve my immediate use case (reserving ports within a test run). I think what you're looking for is a "test run ended" event. That does not imply the shared worker has to shut down (e.g. with watch mode) but resources for that run could be cleaned up. Unless you'd rather have a "teardown of the shared worker because AVA is exiting" kind of event? Let me know and I can open an issue with some API ideas.



> > Tests can wait until their page has been loaded.
> 
> Indeed, and they're doing that in my version as well. I don't think that's the concern?

No, my bad, I was trying to share (poorly) my approach to having Puppeteer load pages. Less RPC, more message driven. But that's just my personal taste.

**Reply by timdp** [reply]
Communication is hard! I feel your pain.

Yep, what I'm after is more of the latter. My intent was/is to keep a single instance of every Shared Worker live for the duration of all tests - a singleton that lives alongside AVA itself, if you will.

So indeed, I need two async hooks per worker: one that runs before all test files, and one that runs after them. The first one is already available today (not wrapped in a hook, but that's not a problem). The second one requires a lot of boilerplate and as we discovered, isn't available in serial mode. I also think it might fail in contrived concurrent scenarios that include brief moments where there are no tests running but that's highly speculative.

Happy to try and contribute something like this myself as well. I'm mostly unfamiliar with the codebase though so it'd probably take a few iterations.

**Reply by novemberborn** [reply]
I think a generic teardown would work for your use case, if you're happy to keep the child processes alive in watch mode. Happy to help if you get stuck.

**Reply by timdp** [reply]
Just to make sure I understand, the intent would be this:

- The default behavior (non-watch mode, optionally serial) would be to keep the shared worker (and e.g. Puppeteer as a result) alive for the duration of all tests, and gracefully terminate it afterwards.
- In watch mode, the shared worker would stay active until I send Ctrl+C (with a timeout, and another Ctrl+C to forcibly terminate).

This seems to align with what you wrote under #2734 (at least the parts I understand without digging deeper) and it makes perfect sense to me.

I appreciate that it's a big refactor though so I'm open to workarounds. Unfortunately, the best I can come up with would be to sidestep shared workers entirely, and instead fork my own thread or daemon process that wraps Puppeteer and includes some keepalive logic - i.e., if it stays idle for a few seconds, assume no new tests are going to come in, and gracefully terminate. However, that would be heavily redundant if AVA can do all the heavy lifting for me.

**Reply by novemberborn** [reply]
Maybe it's OK if Puppeteer is closed, and you start it up again if another test shows up?

**Reply by timdp** [reply]
Is that about the last paragraph or about the intended behavior for AVA? In both cases, I'd argue for a deterministic code path where all tests are handled by the same Puppeteer instance.

Conversely, it'd be even more robust if all tests were handled by their own Puppeteer instance, but since that's too slow, I'd rather have the complete opposite than something in between. For what it's worth, I can also imagine cases where we want to clear Puppeteer's session between tests, but that's completely out of scope here. I do have some ideas around it, and they're compatible with the behavior I'm proposing.

**Reply by novemberborn** [reply]
Sure, just saying, in the meantime you can achieve closing Puppeteer when there are no connected test workers, which assuming concurrent execution is at least better than having each test file spin up its own instance.

**Reply by timdp** [reply]
True, that's easier than maintaining a separate process. I've introduced a `Recyclable` class that's essentially a cyclic state machine in https://github.com/timdp/ava-shared-workers-experiment/commit/9881aff7509dcd3354b715013e7ff8abfe55a9fb. Not exactly pleasant but it's a decent workaround for now. Thanks for the tip!

**Comment by cspotcode** [comment]
Is there a shared worker to run a setup operation once, with all test files waiting for it to complete before running their tests?  This task configures something in the host environment.

`@ava/cooperate` only gets you part of the way there.

An un-managed semaphore can ensure that only one worker does the setup task.  A lock ensures all other workers wait for it to complete.

The missing piece is error reporting.  If the setup task fails, all workers should error.  This could be as simple as:

```typescript
import {setValue, getValue} from @ava/shared-values';
const {errorMessage} = getValue('setup-task-result');
```

**Reply by novemberborn** [reply]
No, but that'd be great to have. Some of the points here are relevant https://github.com/avajs/ava/discussions/3004:

>* A way to load the worker when AVA starts up. This could be declarative in the configuration, or perhaps we can use the factory style of the ava.config.* files and provide a registerSharedWorker() function that follows ava/plugin
>    [ava/plugin.d.ts](https://github.com/avajs/ava/blob/ae0042c95d2dbe72b4910931232ea15f6071e04e/plugin.d.ts#L76)
> * A way so that test runs don't start before these workers are ready.

**Reply by cspotcode** [reply]
I implemented a plugin to do this: https://github.com/cspotcode/ava-shared-setup
Unfortunately, I can't use it myself because I'm stuck using child processes instead of worker_threads.

It implements:
a) a shared key-value store
b) a way to execute one-time setup tasks

Importantly, a setup task does *not* run inside the shared worker!  Rather, it runs in the same thread as your tests, but only in one of them.  All other test threads wait for it to finish and throw the same error message if a failure happens.

This means setup code lives alongside your tests, and running ava is always sufficient to execute setup tasks.  No need to `some-setup-command && ava` in the shell.  IDE integrations should work.  And when running a subset of tests that *don't* require this task to run, it can be skipped.

Usage looks like this:

```
import {Context} from 'ava-shared-setup';
const ctx = new Context('my-context-id');
test('foo', async () => {
  await ctx.task('my-task-id', async () => {
    await exec('start-shared-server');
  });
});
```

**Reply by novemberborn** [reply]
Very cool!
